installers/AppVeyor.hs

```diff
@@ -0,0 +1,160 @@
+import GHC.Generics
+
+-- | Gets CardanoSL.zip corresponding to the src json revision from AppVeyor CI
+downloadCardanoSL :: FilePath -> IO L8.ByteString
+downloadCardanoSL srcJson = do
+  src@CardanoSource{..} <- getCardanoRev srcJson
+  printf ("Fetching GitHub CI status for commit "%n%" of "%n%"/"%n%"\n") srcRev srcOwner srcRepo
+  buildURL <- appVeyorURL src
+  printf ("Build URL is "%s%"\n") (getUrl buildURL)
+  jobs <- appVeyorJobs buildURL
+  printf ("Build has "%d%" job(s), status "%ss%".\n") (length jobs)
+    (map (show . jobStatus) jobs)
+  arts <- appVeyorArtifacts $ map jobId jobs
+  printf ("Artifacts are: "%ss%"\n") (map fst arts)
+  case findAppVeyorArtifact "CardanoSL.zip" arts of
+    Just bs -> printf "Done\n" >> pure bs
+    Nothing -> throwM (MissingArtifactsError $ getUrl buildURL)
+
+n :: Format r (Name a -> r)
+n = makeFormat untagName
+
+ss :: Format r ([Text] -> r)
+ss = makeFormat (T.intercalate ",")
+
+----------------------------------------------------------------------------
+
+-- | Given a build status URL, query AppVeyor for job IDs.
+appVeyorJobs :: GH.URL -> IO [Job]
+appVeyorJobs buildURL = do
+  let Just req = buildAPI buildURL
+  r <- asValue =<< get (show req)
+  pure (r ^.. responseBody . buildJobs)
+
+-- | Download all artifacts for all job ids.
+appVeyorArtifacts :: [Text] -> IO [(Text, L8.ByteString)]
+appVeyorArtifacts = fmap concat . mapM artifactForJobId
+
+-- | Filter the artifact we are interested in.
+findAppVeyorArtifact :: Text -> [(Text, L8.ByteString)] -> Maybe L8.ByteString
+findAppVeyorArtifact name arts = head [bs | (name', bs) <- arts, name' == name]
+
+-- | Downloads all artifacts for a job.
+-- Returns the filename and contents.
+artifactForJobId :: Text -> IO [(Text, L8.ByteString)]
+artifactForJobId jobId = do
+  let req = jobArtifactsURI jobId
+  r <- asValue =<< get req
+  forM (r ^.. responseBody . artifactFilenames) $ \filename -> do
+    let req' = req ++ ('/':T.unpack filename)
+    printf ("Downloading "%w%" ... ") req'
+    r' <- get req'
+    printf "done\n"
+    pure (filename, r' ^. responseBody)
+
+-- | Looks into AppVeyor API build info json to get job IDs
+buildJobs :: Monoid r => Getting r Value Job
+buildJobs = key "build" . key "jobs" . _Array . traverse . _JSON
+
+-- | Looks into AppVeyor API job artifacts json to get the filenames
+artifactFilenames :: Monoid r => Getting r Value Text
+artifactFilenames = _Array . traverse . key "fileName" . _String
+
+-- | The AppVeyor API URL of an build, given its web URL.
+buildAPI :: GH.URL -> Maybe URI
+buildAPI = parseAbsoluteURI . T.unpack . T.replace "/project/" "/api/projects/" . getUrl
+
+-- | The AppVeyor API URL to list artifacts belonging to a job.
+jobArtifactsURI :: Text -> String
+jobArtifactsURI jobId = "https://ci.appveyor.com/api/buildJobs/" <> T.unpack jobId <> "/artifacts"
+
+-- | Use GitHub API to find AppVeyor build status
+appVeyorURL :: CardanoSource -> IO GH.URL
```

Comment:
I'm worried we're getting a proliferation of differently-typed URI values..  I don't have a particular suggestion here -- just a thought that we might want to come up with a kind of a generic approach to the problem.

Reply:
Pretty much unavoidable if you have the URL type from GitHub which is just a newtype Text wrapper. I just went with the way which involved minimal URL parsing.